contracts/0.8.9/WithdrawalQueueBase.sol

```diff
@@ -0,0 +1,524 @@
+        uint256 _ethBudget,
+        uint256 _shareRate,
+        uint256 _startId,
+        uint256 _endId
+    ) public view returns (uint256 finalizableRequestId) {
+        if (_ethBudget == 0) revert ZeroAmountOfETH();
+        if (_shareRate == 0) revert ZeroShareRate();
+        if (_startId <= getLastFinalizedRequestId()) revert InvalidRequestIdRange(_startId, _endId);
+        if (_endId > getLastRequestId()) revert InvalidRequestIdRange(_startId, _endId);
+
+        if (_startId > _endId) return 0; // we have an empty range to search in
+
+        uint256 startRequestId = _startId;
+        uint256 endRequestId = _endId;
+
+        finalizableRequestId = 0;
+
+        while (startRequestId <= endRequestId) {
+            uint256 midRequestId = (endRequestId + startRequestId) / 2;
+            (uint256 requiredEth,) = finalizationBatch(midRequestId, _shareRate);
+
+            if (requiredEth <= _ethBudget) {
+                finalizableRequestId = midRequestId;
+
+                // Ignore left half
+                startRequestId = midRequestId + 1;
+            } else {
+                // Ignore right half
+                endRequestId = midRequestId - 1;
+            }
+        }
+    }
+
+    /// @notice Returns last `requestId` finalizable under given conditions
+    /// @param _ethBudget max amount of eth to be used for finalization
+    /// @param _shareRate share rate that will be applied to requests
+    /// @param _maxTimestamp timestamp that requests should be created before
+    ///
+    /// @dev WARNING! OOG is possible if used onchain, contains unbounded loop inside
+    /// @return finalizableRequestId or 0, if there are no requests finalizable under given conditions
+    function findLastFinalizableRequestId(uint256 _ethBudget, uint256 _shareRate, uint256 _maxTimestamp)
+        public
+        view
+        returns (uint256 finalizableRequestId)
+    {
+        uint256 firstUnfinalizedRequestId = getLastFinalizedRequestId() + 1;
+        finalizableRequestId =
+            findLastFinalizableRequestIdByBudget(_ethBudget, _shareRate, firstUnfinalizedRequestId, getLastRequestId());
+        return findLastFinalizableRequestIdByTimestamp(_maxTimestamp, firstUnfinalizedRequestId, finalizableRequestId);
+    }
+
+    /// @notice Calculates the amount of ETH required to finalize the batch of requests in the queue up to
+    /// `_nextFinalizedRequestId` with given `_shareRate` and the amount of shares that should be burned after
+    /// @param _nextFinalizedRequestId id of the ending request in the finalization batch (>0 and <=lastRequestId)
+    /// @param _shareRate share rate that will be used to calculate the batch (1e27 precision, >0)
+    ///
+    /// @return ethToLock amount of ETH required to finalize the batch
+    /// @return sharesToBurn amount of shares that should be burned after the finalization
+    function finalizationBatch(uint256 _nextFinalizedRequestId, uint256 _shareRate)
```

Comment:
NB: _nextFinalizedRequestId should be strict grater than `getLastFinalizedRequestId()`